sparse/_common.py

```diff
@@ -1211,6 +1212,142 @@ def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
     return _dot_ndarray_coo
 
 
+def _einsum_single(lhs, rhs, operand):
+    """Perform a single term einsum, i.e. any combination of transposes, sums
+    and traces of dimensions.
+
+    Parameters
+    ----------
+    lhs : str
+        The indices of the input array.
+    rhs : str
+        The indices of the output array.
+    operand : SparseArray
+        The array to perform the einsum on.
+
+    Returns
+    -------
+    output : SparseArray
+    """
+    operand = asCOO(operand)
+
+    if lhs == rhs:
+        # XXX: take copy?
+        return operand
+
+    # check if repeated / 'trace' indices mean we are only taking a subset
+    where = {}
+    for i, ix in enumerate(lhs):
+        where.setdefault(ix, []).append(i)
+
+    selector = None
+    for ix, locs in where.items():
+        loc0, *rlocs = locs
+        if rlocs:
+            # repeated index - only select data where all indices match
+            loc0, *locs = locs
+            subselector = (operand.coords[loc0] == operand.coords[rlocs]).all(axis=0)
+            if selector is None:
+                selector = subselector
+            else:
+                selector &= subselector
+
+    # indices that are removed (i.e. not in the output / `perm`)
+    # are handled by `has_duplicates=True` below
+    perm = [lhs.index(ix) for ix in rhs]
+    new_shape = tuple(operand.shape[i] for i in perm)
+
+    # select the new COO data
+    if selector is not None:
+        new_coords = operand.coords[:, selector][perm]
+        new_data = operand.data[selector]
+    else:
+        new_coords = operand.coords[perm]
+        new_data = operand.data
+
+    if not rhs:
+        # scalar output - match numpy behaviour by not wrapping as array
+        return new_data.sum()
+
+    return operand.__class__(new_coords, new_data, shape=new_shape, has_duplicates=True)
```

Comment:
Why not use `COO` directly?

Reply:
It was just to avoid the circular import error of putting `from .coo import COO` in `_common.py`,but suggestions welcome.

Reply:
Well, operand may not be `COO`. I'd instead import inside the function.